### PR TITLE
Localized builds

### PR DESCRIPTION
**Pull request by cdmihai**
- localized builds are enabled by setting the property `LocalizedBuild` to true
- localization is done via Satellite Assemblies
- localized resx files are generated on the fly in the intermediary output and not under version control
- loc testing can be done by setting `LocalizedTestBuild` to true. This replaces all strings with `!{resource id}!{English resource}!{Translated resource}!`
   - setting `LocalizedTestString` to some value replaces all resource strings with that value
- CI does localized builds by default

Future work. I suggest doing these in future PRs:
- xlab machines have a weird codepage set for their console (437, which is very USish). Because of that, [this method](https://github.com/Microsoft/msbuild/blob/f90dcff1db3b10c3b136bed1a23ce710f7c579cd/src/XMakeCommandLine/XMake.cs#L1216-L1247) resets the UI culture to en-US, which causes the English satellites to be picked up. I do not know whether normal windows installations have the same weird codepage. Contacted the world ready folks to ask them about that method, and what we should with it, if anything. To get localized satellites picked on xlab machines the workaround is doing `chcp 65001`
- migrate xlf sync tasks to xliffparser repo
- update our dependency to xliffparser when a new package is available, so we can use the `state=updated` xlf note for updated strings (as opposed to what happens now, where both new and updated resources get stamped with a "new" tag).

Todo before merging:
- [x] Write wiki on localization process and interaction with the localization team

### CODE REVIEW

**Comment by cdmihai** [comment]
@dotnet-bot test this please

File: netci.groovy [hunk 1]
```diff
@@ -29,7 +29,7 @@ def project = GithubProject
             case 'Windows_NT':
                 newJob.with{
                     steps{
-                        batchFile("call \"C:\\Program Files (x86)\\Microsoft Visual Studio 14.0\\VC\\vcvarsall.bat\" && RebuildWithLocalMSBuild.cmd")
+                        batchFile("call \"C:\\Program Files (x86)\\Microsoft Visual Studio 14.0\\VC\\vcvarsall.bat\" && RebuildWithLocalMSBuild.cmd /p:LocalizedBuild=true")
```
Comment (rainersigwald) [inline netci.groovy]: Changing the script to take arguments is good, but I think I'd rather see the default be localized (with an opt-out). Not urgent though.

File: src/UpdateLocalizedResources.targets [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+      Condition="%(Identity) != ''"
+      ResxPath="%(ChildResx)"
+      XlfPath="@(XlfFiles)"
+      ReplaceWithTestString="$(LocalizedTestBuild)"
+      TestString="$(LocalizedTestString)"
+    />
+
+  </Target>
+
+    <!-- Copy paste and modify OpenSourceSign from sign.targets because it only signes the main assembly
```
Comment (rainersigwald) [inline src/UpdateLocalizedResources.targets]: typo: `signes`

File: src/UpdateLocalizedResources.targets [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+      ReplaceWithTestString="$(LocalizedTestBuild)"
+      TestString="$(LocalizedTestString)"
+    />
+
+  </Target>
+
+    <!-- Copy paste and modify OpenSourceSign from sign.targets because it only signes the main assembly 
+  and provides no extension mechanism to add more assemblies to it
+  
+  If satellite assemblies do not get their delay sign bit switched off, the main assemblies won't load them
```
Comment (rainersigwald) [inline src/UpdateLocalizedResources.targets]: obligatory "have you sent a PR upstream?"

File: build/LocalizationTasks/Microsoft.Build.LocalizationTasks.csproj [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    <Compile Include="*.cs"/>
+  </ItemGroup>
+
+  <ItemGroup>
+    <Reference Include="System"/>
+    <Reference Include="System.Text.RegularExpressions"/>
+    <Reference Include="Microsoft.Build.Framework" />
+    <Reference Include="Microsoft.Build.Utilities.Core" />
+    <Reference Include="fmdev.XliffParser">
+      <HintPath>$(XliffParserDir)fmdev.XliffParser.dll</HintPath>
```
Comment (rainersigwald) [inline build/LocalizationTasks/Microsoft.Build.LocalizationTasks.csproj]: Can you not get this reference via NuGet?

File: src/Utilities/Microsoft.Build.Utilities.csproj [hunk 5]
```diff
@@ -209,15 +209,20 @@
     </Compile>
     <!-- Win32 RC Files -->
     <RCResourceFile Include="native.rc" />
+
```
Comment (rainersigwald) [inline src/Utilities/Microsoft.Build.Utilities.csproj]: delete spaces :)

File: src/XMakeBuildEngine/Resources/Strings.resx [hunk 6]
```diff
@@ -1193,10 +1193,6 @@
     <value>MSB4207: Only an in-process affinity may be specified for projects which have registered host objects.</value>
     <comment>{StrBegin="MSB4207: "}</comment>
   </data>
-  <data name="InvalidAffinityForProjectWithHostObject" UESanitized="false" Visibility="Public">
-    <value>MSB4208: Only an in-process affinity may be specified for projects which have registered host objects.</value>
-    <comment>{StrBegin="MSB4208: "}</comment>
-  </data>
```
Comment (rainersigwald) [inline src/XMakeBuildEngine/Resources/Strings.resx]: can you make removing this (redundant?) string its own commit?

File: src/XMakeCommandLine/UnitTests/XMake_Tests.cs [hunk 7]
```diff
@@ -1783,11 +1783,32 @@ private string CopyMSBuild()
...
                 }
 
+                foreach (var directory in directoriesToCopy)
+                {
+                    foreach (var sourceFile in Directory.EnumerateFiles(directory, "*"))
+                    {
+                        var destinationFile = sourceFile.Replace(source, dest);
+
+                        var directoryName = Path.GetDirectoryName(destinationFile);
+                        if (!Directory.Exists(directoryName))
```
Comment (rainersigwald) [inline src/XMakeCommandLine/UnitTests/XMake_Tests.cs]: The usual wisdom is that this is a race condition (the directory could get deleted between this line and the next), so there's no point in checking first--just call `CreateDirectory`.

File: src/XMakeTasks/System.Design.cs [hunk 8]
```diff
@@ -91,11 +91,24 @@ internal sealed class SR
         internal const string InvalidIdentifier = "InvalidIdentifier";
 
         private static SR s_loader = null;
-        private ResourceManager _resources;
+        private MainAssemblyFallbackResourceManager _resources;
+
+        /// <summary>
+        /// The containing assembly is set to lookup up resources for the neutral language in satellite assemblies, not in the main assembly.
```
Comment (rainersigwald) [inline src/XMakeTasks/System.Design.cs]: typo: `lookup up`

File: src/XMakeTasks/Strings.resx [hunk 9]
```diff
@@ -1562,12 +1562,6 @@
         LOCALIZATION: Please don't localize "CopyLocal" this is an item meta-data name.
    </comment>
   </data>
-  <data name="ResolveAssemblyReference.NoBecauseParentReferencesFoundInGac">
-    <value>This reference is not "CopyLocal" because the CopyLocalDependenciesWhenParentReferenceInGac property is set to false and all the parent references for this reference are found in the GAC.</value>
-    <comment>
-        LOCALIZATION: Please don't localize "CopyLocal" this is an item meta-data name. "CopyLocalDependenciesWhenParentReferenceInGac" is a property name.
-   </comment>
-  </data>
```
Comment (rainersigwald) [inline src/XMakeTasks/Strings.resx]: you can put this one with the other remove-redundant thing

File: src/UpdateLocalizedResources.targets [hunk 10]
```diff
@@ -0,0 +1,108 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Project DefaultTargets="UpdateXlf" ToolsVersion="12.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  
+  <UsingTask AssemblyFile="$(BuildBinDir)\Microsoft.Build.LocalizationTasks.dll" Condition="$(LocalizedBuild) == 'true'" TaskName="SaveXlfToResx"/>
+  <UsingTask AssemblyFile="$(BuildBinDir)\Microsoft.Build.LocalizationTasks.dll" Condition="$(LocalizedBuild) == 'true'" TaskName="UpdateXlfFromResx"/>
+  <UsingTask AssemblyFile="$(BuildBinDir)\Microsoft.Build.LocalizationTasks.dll" Condition="$(LocalizedBuild) == 'true'" TaskName="EmitLocalizedResources"/>
+
+  <!-- Emit the localized EmbeddedResource items before the common targets start using them -->
+  <PropertyGroup>
+    <PrepareResourceNamesDependsOn>
```
Comment (rainersigwald) [inline src/UpdateLocalizedResources.targets]: I'm not a fan of this mechanism of injecting your process. I prefer using `BeforeTargets` on the targets you're injecting.

File: src/UpdateLocalizedResources.targets [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+      LocalizedResxRoot="$(IntermediateOutputPath)"
+      RelativePathToXlfRoot=".\xlf\">
+
+      <Output TaskParameter="ResolvedXlfResources" ItemName="Xlf_Temp" />
+      <Output TaskParameter="ResolvedLocalizedResxResources" ItemName="Resx_temp" />
+
+    </EmitLocalizedResources>
+
+    <ItemGroup>
+      <XlfFiles Include="@(Xlf_Temp)"/>
```
Comment (rainersigwald) [inline src/UpdateLocalizedResources.targets]: Why these temp items?

File: src/UpdateLocalizedResources.targets [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+      <EmbeddedResource Include="@(Resx_temp)"/>
+    </ItemGroup>
+
+  </Target>
+
+    <!-- The default english resources do not have corresponding xlf files. Simply copy them and add en in the name -->
+    <!-- TODO: Maybe create a dummy xlf file for English so we can delete this target? -->
```
Comment (rainersigwald) [inline src/UpdateLocalizedResources.targets]: This seems like a fine idea. Is there a downside?

File: src/UpdateLocalizedResources.targets [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+  <Target Name="UpdateLocalizedResxForEnglishResx">
+
+      <ItemGroup>
+        <EnglishResources Include="@(NeutralResources -> '$(IntermediateOutputPath)%(Filename).en%(Extension)')"/>
```
Comment (rainersigwald) [inline src/UpdateLocalizedResources.targets]: Could you do this transform in the task directly to avoid a temporary Item?

File: src/UpdateLocalizedResources.targets [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+  <!-- Emit the localized EmbeddedResource items before the common targets start using them -->
+  <PropertyGroup>
+    <PrepareResourceNamesDependsOn>
+      EmitLocalizedResources;
+      UpdateLocalizedResxForEnglishResx;
+      $(PrepareResourceNamesDependsOn)
+    </PrepareResourceNamesDependsOn>
+  </PropertyGroup>
+
+  <Target Name="EmitLocalizedResources">
```
Comment (rainersigwald) [inline src/UpdateLocalizedResources.targets]: Targets written by an MSBuild core team member that don't have incremental behavior? :scream:

(this can wait, but there should at least be a bug!)

File: build/LocalizationTasks/EmitLocalizedResources.cs [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+
+            xlf.SetMetadata(ChildResxMetadata, resxPath);
+
+            return resxItem;
+        }
+
+        private IEnumerable<ITaskItem> ComputeXlfResources(ITaskItem neutralResource)
+        {
+            var neutralResxRootDirectory = Path.GetDirectoryName(neutralResource.GetMetadata("FullPath"));
+            var xlfRootPath = Path.Combine(neutralResxRootDirectory, RelativePathToXlfRoot);
```
Comment (rainersigwald) [inline build/LocalizationTasks/EmitLocalizedResources.cs]: Why prefer this to just doing the computation in msbuild and using it directly here?

File: build/LocalizationTasks/EmitLocalizedResources.cs [hunk 16]
```diff
@@ -0,0 +1,112 @@
...
+
+            // todo: throw if xlf root does not exist; throw in other places too
```
Comment (rainersigwald) [inline build/LocalizationTasks/EmitLocalizedResources.cs]: sounds good :)

File: build/LocalizationTasks/EmitLocalizedResources.cs [hunk 17]
```diff
@@ -0,0 +1,112 @@
...
+            {
+                Log.LogMessage("No neutral resources found");
+                return true;
+            }
+
+            ResolvedXlfResources = NeutralResources.SelectMany(ComputeXlfResources).ToArray();
+
+            ResolvedLocalizedResxResources = ResolvedXlfResources.Select(ComputeLocalizedResource).ToArray();
+
+            return true;
```
Comment (rainersigwald) [inline build/LocalizationTasks/EmitLocalizedResources.cs]: MSBuild idiom: always return `!Log.HasLoggedErrors`--this prevents situations where you fail the build (`return false`) but never logged an error (so there's nothing to go on when it happens and your build fails).

File: build/LocalizationTasks/EmitLocalizedResources.cs [hunk 18]
```diff
@@ -0,0 +1,112 @@
...
+            var xlfRootPath = Path.Combine(neutralResxRootDirectory, RelativePathToXlfRoot);
+
+            // todo: throw if xlf root does not exist; throw in other places too
+
+            var resx = neutralResource.ItemSpec;
+
+            return (Directory.EnumerateFiles(xlfRootPath).Where(f => IsValidXlf(f, resx)).Select((f => CreateXLFTaskItemForNeutralResx(f, resx))));            
+        }
+
+        private bool IsValidXlf(string xlfPath, string resx)
```
Comment (rainersigwald) [inline build/LocalizationTasks/EmitLocalizedResources.cs]: Why is this needed? It seems a bit scary.

Actually the whole enumerate-files approach scares me a bit; I prefer an explicit-inputs approach.

File: build/LocalizationTasks/SaveXlfToResx.cs [hunk 19]
```diff
@@ -0,0 +1,64 @@
...
+                if (!string.IsNullOrEmpty(ReplaceWithTestString) && ReplaceWithTestString.Equals("true"))
+                {
+                    ReplaceResourceValuesWithTestString(xlfDocument);
+                }
+
+                xlfDocument.SaveAsResX(ResxPath);
+
+                return true;
+            }
+            catch (Exception e)
```
Comment (rainersigwald) [inline build/LocalizationTasks/SaveXlfToResx.cs]: can you be more specific than just `Exception`?

File: build/LocalizationTasks/SaveXlfToResx.cs [hunk 20]
```diff
@@ -0,0 +1,64 @@
...
+        public string XlfPath { get; set; }
+
+        /// <summary>
+        /// When set to "true" it makes the task replace the resource values with another value.
+        /// Useful for testing localized builds.
+        /// 
+        /// If <see cref="TestString"/> is non empty, resource values get replaced with it.
+        /// If it is empty, resource values get replaced with !resource_id!english_resource!localized_resource!
+        /// </summary>
+        public string ReplaceWithTestString { get; set; }
```
Comment (rainersigwald) [inline build/LocalizationTasks/SaveXlfToResx.cs]: This seems pretty cool, but is it so strongly coupled with this functionality that it belongs here?

File: netci.groovy (hunk 1 above)
Comment (cdmihai) [inline netci.groovy]: Thought about that, but it means whenever you run RebuildWithLocalMSBuild on your local machine, you might get dirty xlf files if any resource changed. This might lead to people committing xlf file changes while the loc team is working on them as well.

File: src/UpdateLocalizedResources.targets (hunk 3 above)
Comment (cdmihai) [inline src/UpdateLocalizedResources.targets]: Shamefully, no. I'll add it to my todo list, along with moving the tasks to xliff parser.

File: netci.groovy (hunk 1 above)
Comment (rainersigwald) [inline netci.groovy]: Good point; I find that convincing.

File: build/LocalizationTasks/Microsoft.Build.LocalizationTasks.csproj (hunk 4 above)
Comment (cdmihai) [inline build/LocalizationTasks/Microsoft.Build.LocalizationTasks.csproj]: I had forgotten csproj and project.json actually work together :)

File: src/XMakeCommandLine/UnitTests/XMake_Tests.cs (hunk 7 above)
Comment (cdmihai) [inline src/XMakeCommandLine/UnitTests/XMake_Tests.cs]: True, that is indeed wise. :)

File: src/UpdateLocalizedResources.targets (hunk 10 above)
Comment (cdmihai) [inline src/UpdateLocalizedResources.targets]: I also had a philosophical debate with myself about this one. Can I do `BeforeTargets="$(PrepareResourceNamesDependsOn)"`? I really want these two targets (that inject and create resource items and files) to be the very first, since most likely all the other targets in that property will depend on them.

Also, I want to reference the property and not peek inside it to see what's the first target there, since that feels a bit like breaking encapsulation, and other projects could preppend other targets in that string, making my first choice not "the first one" anymore.

File: src/UpdateLocalizedResources.targets (hunk 12 above)
Comment (cdmihai) [inline src/UpdateLocalizedResources.targets]: Potential money wise issues with translation. I'll go ahead and ask the loc team if it's alright.

File: src/UpdateLocalizedResources.targets (hunk 11 above)
Comment (cdmihai) [inline src/UpdateLocalizedResources.targets]: I didn't know whether task outputs are additive, or whether they overwrite the item type with new contents. Seemed safer and more explicit this way. Do they overwrite, or do they add? :)

File: build/LocalizationTasks/EmitLocalizedResources.cs (hunk 15 above)
Comment (cdmihai) [inline build/LocalizationTasks/EmitLocalizedResources.cs]: I tried writing the msbuild code for it but after 1-2 hours of failed attempts and very dubious looking hacked up property functions, I succumbed to something more elegant, C# :)

File: build/LocalizationTasks/EmitLocalizedResources.cs (hunk 18 above)
Comment (cdmihai) [inline build/LocalizationTasks/EmitLocalizedResources.cs]: For explicit inputs, I would have to pass in another item of allowed language codes (`en;cs;de;ja; etc`) and pre-compute the xlf names from those. But then `en` is special, since there are no xlfs for it, and I'd have to have special code for it. Because of that, I went with automatic discover via IsValidXlf. One alternative is to not include `en` in the item of language codes. But then `en` resources still popup via another target, making things more confusing.

File: build/LocalizationTasks/SaveXlfToResx.cs (hunk 19 above)
Comment (cdmihai) [inline build/LocalizationTasks/SaveXlfToResx.cs]: Good question. My reasoning was on the lines of: "I have no idea what exceptions might get transitively thrown from xliff parser, so then instead of letting a potential exception break out of the task, I'd rather log it as an error and have the task return false".

File: build/LocalizationTasks/SaveXlfToResx.cs (hunk 20 above)
Comment (cdmihai) [inline build/LocalizationTasks/SaveXlfToResx.cs]: Kind of, because I have to edit the in-memory xlf and then ask it to serialize to resx. If I take it out, it will actually be the entire thing and this task would remain a copy but without the testing code in it. 

The alternative is to implement the localization testing functionality in a separate task and have it parse the generated resx and edit that. But that's more work. Do you think the testing functionality should be done at the resx level, via resx resource readers / writers?

File: src/Utilities/Microsoft.Build.Utilities.csproj (hunk 5 above)
Comment (cdmihai) [inline src/Utilities/Microsoft.Build.Utilities.csproj]: I specifically added them in to make the csproj easier to read. I'll take them out if I'm breaking some styling convention :)

File: src/UpdateLocalizedResources.targets (hunk 14 above)
Comment (cdmihai) [inline src/UpdateLocalizedResources.targets]: I had tried but bailed out of the change because:
- it required splitting a target in multiple pieces (some parts needed to output items indifferent of incremental check), making things more ugly
- a task from a target also depends on a property (doing localized test builds), besides files. If I make the target incremental, you have to do a clean build if you want to transform the satellites into test satellites.

File: src/Utilities/Microsoft.Build.Utilities.csproj (hunk 5 above)
Comment (rainersigwald) [inline src/Utilities/Microsoft.Build.Utilities.csproj]: I care (a bit) about the trailing space characters. Bare newlines for paragraph spacing I wholeheartedly approve of.

(I'm cursed by my ongoing "need" to have visible whitespace on, so I see this sort of things.)

File: build/LocalizationTasks/SaveXlfToResx.cs (hunk 20 above)
Comment (rainersigwald) [inline build/LocalizationTasks/SaveXlfToResx.cs]: Yeah, that's reasonable. I was thinking more along the lines of factoring most of this out and having two tasks call into the base code, one with editing. But that's the architecture :rocket: in me. This seems fine.

File: build/LocalizationTasks/SaveXlfToResx.cs (hunk 19 above)
Comment (rainersigwald) [inline build/LocalizationTasks/SaveXlfToResx.cs]: The standard advice, which I'm mostly just parroting here, is that catching `Exception` is too general, because it catches 
* Interesting runtime failures that might be actionable (what you want)
* Runtime failures due to programmer error that are bugs but not actionable for the end user (_maybe_ ok, might be better represented elsewhere)
* Systemic failures due to process or machine state (when you mostly just want to crash)

MSBuild itself already [provides an all-purpose task-exception handler](https://github.com/Microsoft/msbuild/blob/4b30fac51ab58dadfffedb2a57e83b8398e3a5e1/src/XMakeBuildEngine/BackEnd/Components/RequestBuilder/TaskBuilder.cs#L773-L902), so I'd say to just depend on that if there's no `XliffParserException` that's specific to the code that's being called. Plus, it doesn't look like there's [much in the way of exceptions](https://github.com/fmuecke/XliffParser/search?utf8=%E2%9C%93&q=throw) in there, currently.

Short further reading: https://msdn.microsoft.com/en-us/library/ms182137.aspx
LONG but interesting further reading: [Midori error model](http://joeduffyblog.com/2016/02/07/the-error-model/)

File: build/LocalizationTasks/EmitLocalizedResources.cs (hunk 18 above)
Comment (rainersigwald) [inline build/LocalizationTasks/EmitLocalizedResources.cs]: Is this correct?

* The list of languages is canonically determined by files in the repo.
* There's also an implicit `en` resource that needs to be treated differently.
* If you had a list of languages, you'd have to reconstruct file paths from the language (it's not a guaranteed map)?

That might be enough to convince me. But could you just do `<XlfFiles Include="*.xlf" />` or something?

File: src/UpdateLocalizedResources.targets (hunk 14 above)
Comment (rainersigwald) [inline src/UpdateLocalizedResources.targets]: > some parts needed to output items indifferent of incremental check

I'm not sure I understand. Can you elaborate?

> test satellites

That's definitely interesting. Maybe `Touch` an additional output if there you're doing test satellites? That would require a clean build to switch back, still.

File: src/UpdateLocalizedResources.targets (hunk 12 above)
Comment (rainersigwald) [inline src/UpdateLocalizedResources.targets]: I'm convinced. Can you put that reasoning in the comment or commit message?

File: src/UpdateLocalizedResources.targets (hunk 11 above)
Comment (rainersigwald) [inline src/UpdateLocalizedResources.targets]: It's additive, but I'll be darned if I can find that documented anywhere! #805

File: src/UpdateLocalizedResources.targets (hunk 10 above)
Comment (rainersigwald) [inline src/UpdateLocalizedResources.targets]: Yes, you can depend on the property, and I think that makes sense. You might do `BeforeTargets="$(PrepareResourceNamesDependsOn);PrepareResourceNames"` in case it's empty/conditioned out.

File: src/Utilities/Microsoft.Build.Utilities.csproj (hunk 5 above)
Comment (cdmihai) [inline src/Utilities/Microsoft.Build.Utilities.csproj]: okay, gonna scan for spurious white-spaces. I should enable white-space visibility as well :)

File: src/UpdateLocalizedResources.targets (hunk 10 above)
Comment (cdmihai) [inline src/UpdateLocalizedResources.targets]: will do

File: src/UpdateLocalizedResources.targets (hunk 14 above)
Comment (cdmihai) [inline src/UpdateLocalizedResources.targets]: > > some parts needed to output items indifferent of incremental check
> I'm not sure I understand. Can you elaborate?

The following target:
```xml
  <Target Name="UpdateLocalizedResxForEnglishResx">

      <ItemGroup>
        <EnglishResources Include="@(NeutralResources -> '$(IntermediateOutputPath)%(Filename).en%(Extension)')"/>
      </ItemGroup>

      <Copy
        SourceFiles="@(NeutralResources)"
        DestinationFiles="@(EnglishResources)"
      />

      <ItemGroup>
        <EmbeddedResource Include="@(EnglishResources)">
          <LogicalName>$(AssemblyName).%(FileName).resources</LogicalName>
        </EmbeddedResource>
      </ItemGroup>      

  </Target>
```

Both produces the intermediary english resx, and adds them to EmbeddedResource. To make copy incremental, I have to extract the item emitting part to another target that always runs after this one. Or, I can move item emitting code in the previous target that emits other items, but then I have to make sure the incremental english resx target runs before that one, so the file specific transforms to compute the logical name work. OK, now that I talked through it it looks feasible :)

> That's definitely interesting. Maybe Touch an additional output if there you're doing test satellites? That would require a clean build to switch back, still.
Aha, make a file lock as an input that gets touched in a previous target if the non-file state is the right one? I see, clever trick :)

Ok then, nailed both problems. Incremental targets it is :)

File: build/LocalizationTasks/EmitLocalizedResources.cs (hunk 18 above)
Comment (cdmihai) [inline build/LocalizationTasks/EmitLocalizedResources.cs]: That's correct. In my task, the main inputs are the neutral resources. From their localization I compute the root of xlf and then enumerate them. And then from the xlf file names I generate the intermediary localized resx paths from.

The problem with my approach is that if you delete one xlf file, nothing tells the build is invalid (it just enumerates files that match a pattern). So for better correctness it would be nice to also have an explicit list of language codes and throw if the xlf of any is missing.

> But could you just do `<XlfFiles Include="*.xlf" />` or something?
I guess I could do `<XlfFiles Include="@(NeutralResource->'%Root%Directory\xlf\*xlf')" />`. And then compute the intermediary resx off of that.

Comment (cdmihai) [inline build/LocalizationTasks/EmitLocalizedResources.cs]: Actually that would be a one to many transformation. I'll see if msbuild supports that

File: build/LocalizationTasks/SaveXlfToResx.cs (hunk 19 above)
Comment (cdmihai) [inline build/LocalizationTasks/SaveXlfToResx.cs]: New bed-time reading :)

I guess it makes sense overall. I guess we can also beef up documentation for task writers on what's the contract around exceptions and error handling. I remember the msbuild book saying something on the lines of "tasks should avoid throwing exceptions at all costs". Hence I caught everything and transformed it into a log error :)

File: src/XMakeTasks/Strings.resx (hunk 9 above)
Comment (cdmihai) [inline src/XMakeTasks/Strings.resx]: I tried, but bailed after ~15min. Some of the edits are in files that I also moved, making everything a pain. I vote for not splitting that monster commit :)

Comment (rainersigwald) [inline src/XMakeTasks/Strings.resx]: :ok: Thanks for trying :)

**Comment by cdmihai** [comment]
@dotnet-bot test Windows_NT Build please

**Comment by cdmihai** [comment]
Wrote localization documentation: https://github.com/Microsoft/msbuild/wiki/Dealing-With-Localization

**Comment by rainersigwald** [comment]
Documentation 👍 

LGTM.
